### ISSUE TITLE
Bug: Static noise appears over a long period of audio processing

### ISSUE THREAD

**Issue by nodeful** [bitgapp/eqMac#224]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%24100%20Rewarded-%237E24E3.svg" />](https://issuehunt.io/r/bitgapp/eqMac/issues/224)

**Disclaimer:** 
Please go through [existing issues](https://github.com/bitgapp/eqMac/issues?q=is%3Aopen+is%3Aissue+label%3A%22type%3A+bug%22) to avoid creating duplicates. If you found a similar issue please post a comment there. If you are uncertain your issue is related to others then create the issue. Obvious ignorance of this rule might get you blocked from posting new issues. Please tick the box below to show that you have followed the guideline.
- [x] I have checked for a similar issue and sure it hasn't been reported before.

**Describe the bug**
Reported in issues: #36 #79 #152 #174 #180 #187 #199 #210

I (the developer) have never actually seen this problem occur to me, probably due to the specific hardware setup that I have. However, many users have reported and this is a critical issue to fix.

In #180 user @ekarlo have done a lot of work to mitigate the issue with the old eqMac2 release. 
Most likely issue relies in the Sampling Rate mismatch and the slow drift of the Ring Buffer when processing the audio. 

Probably also related to the sync loss and audio latency added reported in: #118 #144 #212 
Need to look into it more.

**Steps to Reproduce**
_Waiting for users to report their hardware setups to achieve a similar setup for reproduction._

**Expected behaviour**
The Audio should stay clear no matter how long eqMac process is running for.

**Setup information:**
_Waiting for feedback from users about the setup that they use._

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $100.00)

- [<img src='https://avatars0.githubusercontent.com/u/8472525?v=4' alt='nodeful' width=24 height=24> nodeful](https://issuehunt.io/u/nodeful) ($100.00)

### Submitted pull Requests
- [#322 Updates For Static Issue](https://issuehunt.io/r/bitgapp/eqMac/pull/322)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/bitgapp/eqMac/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

**Comment by nodeful** [comment]
The bug is likely related to #224 and probably fixing one of these will fix the other.

**Comment by mwmrshll** [comment]
@nodeful, I'm getting this bug running through a Thunderbolt 3 port into a Caldigit TS3+ dock with speakers plugged into its 3.5mm jack. As others said on the previous issue, it goes away temporarily if I switch audio source away and then back to eqMac, and then returns after a couple of minutes.

**Comment by NielsKersic** [comment]
@nodeful I'm running into the same issue with my Revonext QT2S headphones, my KZ ZS10 Pro's and when I use my Bose QC 35 in wired mode. I might be wrong, but it appears that the issue gets worse (appears sooner, progresses quicker) when the bass levels are boosted significantly. I'm running a 2019 15" MacBook Pro.

**Comment by nodeful** [comment]
What Sample rate are you guys using on the output devices? You can find that in the `Audio MIDI Setup.app` on your mac

**Comment by NielsKersic** [comment]
eqMac is at 44.1K the rest is all at 48K

**Comment by nodeful** [comment]
What if you set everything to 44.1K before running eqMac?

**Comment by rome3ro** [comment]
I did that setting up to 44 before installing actually is happening anyway I'm using iTunes on Catalina and if you leave the music running it comes worse the noise

**Comment by rome3ro** [comment]
Switching between output the noise go away immediately however after a while comes again

**Comment by rome3ro** [comment]
Maybe I'm wrong but on eqmac2 I could it switch on the mac (top bar -> sound icon) switch between output with this new version I can do it but after select for example from "eqmac" option to "Macbook Pro Speakers" nothing sounds even if I increase the volume, not sure if this is a bug or not, but as I said on previous version I was able to do that and the sound coming out from the selection

**Comment by NielsKersic** [comment]
@nodeful Tried switching to 44.1K before starting eqMac. Same issue still appeared after a while, although the crackling seemed to be of a higher frequency, although that might be unrelated. Let me know if there's anything else I can try or if there's any info that would help you.

**Comment by mwmrshll** [comment]
Thanks for the response. I switched the speaker to 44.1k with Audio MIDI Setup. This seemed to help at first, but the problem came back (maybe it gained me an extra minute or so). I've noticed that eqMac is set at 44.1k (or 48k) 2 ch 32-bit float, while my speakers can only be set to 44.1k/48k 2 ch 16/24-bit integer. So could the problem arise as eqMac is trying to convert the float sound stream to int? (forgive me if I'm nowhere in the ballpark, not really my area!)

**Comment by mwmrshll** [comment]
Furthermore, when listening through my Sony WH-1000XM3 headphones, there is no static build up. The headphones' format is 48k 2 ch 32-bit float. Even with eqMac set to 44.1k 2 ch 32-bit float, it works fine. So the issue only seems to arise with the float/int difference.

**Comment by kimbeejay** [comment]
Same here as @NielsKersic has described

**Comment by isdaves** [comment]
Experiencing the same symptoms on a MBP 15" 2018 on Mojave, using the integrated speakers of my external LG display (connected via HDMI to Thunderbold3 port)
* Slight latency increase
* Crackling

Output devices:
* eqMac: 2ch 32bit float - 44.1Khz
* LG: 2ch 24bit integer - 48Khz

As others mentioned, it goes away if I switch sources. Also, this never happened to me with eqMac2, which I was using up until now.

**Comment by nodeful** [comment]
I have just release a new version v0.1.0 where I tried to mitigate the issue.
Could you please download the latest version from https://eqmac.app or GitHub, or update directly from the app. Test and report back!

**Comment by jeffstampes** [comment]
I happened to check in and saw that release a few minutes after you released it.   I've been running a couple of hours on a setup that always turned to static within 90 minutes, so I'll give you a preliminary report that it's at least staying stable longer.   Thanks for looking at it!

**Comment by MrSpethial** [comment]
I've tried installing the new version, I'm now getting no audio output at all (While the app is running).

Closing it audio works fine.

2018 Macbook pro 13 -> Inbuilt speakers -> Spotify.

**Comment by isdaves** [comment]
The new one crashes for me upon launching it

```
[04/29 13:33:36 @ ~] /Applications/eqMac.app/Contents/MacOS/eqMac
2020-04-29 13:33:39.758 eqMac[19939:62538] Sentry Started -- Version: 4.5.0
dyld: lazy symbol binding failed: Symbol not found: _objc_opt_class
  Referenced from: /Applications/eqMac.app/Contents/MacOS/../Frameworks/WebViewJavascriptBridge.framework/Versions/A/WebViewJavascriptBridge (which was built for Mac OS X 10.15)
  Expected in: /usr/lib/libobjc.A.dylib

dyld: Symbol not found: _objc_opt_class
  Referenced from: /Applications/eqMac.app/Contents/MacOS/../Frameworks/WebViewJavascriptBridge.framework/Versions/A/WebViewJavascriptBridge (which was built for Mac OS X 10.15)
  Expected in: /usr/lib/libobjc.A.dylib

ERROR: SentryCrashCPU.c (58): _Bool sentrycrashcpu_i_fillState(const thread_t, const thread_state_t, const thread_state_flavor_t, const mach_msg_type_number_t): thread_get_state: (ipc/send) invalid destination port
ERROR: SentryCrashCPU.c (58): _Bool sentrycrashcpu_i_fillState(const thread_t, const thread_state_t, const thread_state_flavor_t, const mach_msg_type_number_t): thread_get_state: (ipc/send) invalid destination port
Abort trap: 6
```

Tried running the previous v0.0.1 (reported as v4.4.0 in CLI) and it also crashes now :(

```
[04/29 13:35:22 @ ~] /Applications/eqMac.app/Contents/MacOS/eqMac
2020-04-29 13:35:24.452 eqMac[20571:64943] Sentry Started -- Version: 4.4.0
ERROR: SentryCrashCPU.c (58): _Bool sentrycrashcpu_i_fillState(const thread_t, const thread_state_t, const thread_state_flavor_t, const mach_msg_type_number_t): thread_get_state: (ipc/send) invalid destination port
ERROR: SentryCrashCPU.c (58): _Bool sentrycrashcpu_i_fillState(const thread_t, const thread_state_t, const thread_state_flavor_t, const mach_msg_type_number_t): thread_get_state: (ipc/send) invalid destination port
2020-04-29 13:35:25.138 eqMac[20571:64943] *** Terminating app due to uncaught exception 'com.apple.coreaudio.avfaudio', reason: 'required condition is false: format.sampleRate == hwFormat.sampleRate'
*** First throw call stack:
<snip>
)
libc++abi.dylib: terminating with uncaught exception of type NSException
Abort trap: 6
```

**Comment by rome3ro** [comment]
> I've tried installing the new version, I'm now getting no audio output at all (While the app is running).
> 
> Audio works fine when EQ is closed.
> 
> 2018 Macbook pro 13 -> Inbuilt speakers -> Spotify.

Check on Audio Midi Setup on Mac that the other audios has 44.1 k configured instead 48

**Comment by kesarion** [comment]
Also crashes for me, MacOS 10.14.6. Old version as well (0.0.1). I tried removing all traces of the app and the driver, but still nothing on reinstall.

**Comment by nodeful** [comment]
@kesarion so before the update, the old build 0.0.1 was fine and now after the update even it is crashing? 🤔

**Comment by nodeful** [comment]
Btw the new Build should have no issues handling any sample rate. The driver will adapt to any Sample Rate set on your output device. The crashes after the update are related to some other issues not connected to this one.

**Comment by kesarion** [comment]
@nodeful No, it started happening with 0.0.1. I reinstalled it multiple times (along with the driver), because I was having this buzzing issue. At some point I thought I'd try one more time after setting the speaker output bit rate to 44.1 kHz and it didn't open anymore after that, not even after a couple of reinstalls, not even after setting the output back to 48.0 kHz. Something permanent happened and I don't know what, nor how to reset it without reinstalling the system.

The difference with the new 0.1.0 is that MacOS actually tells me the app crashed with the Send Report message; before, it would crash silently, no message, it wouldn't even get there.

**Comment by nodeful** [comment]
The buzzing and app crashing can't be related at all. Please send crash logs so I can point you to the right issue @kesarion

**Comment by nodeful** [comment]
Whoever has problems running v0.1.0 please download the pre-release of v0.1.1 here: https://github.com/bitgapp/eqMac/releases/tag/v0.1.1

**Comment by kesarion** [comment]
@nodeful Wow, it's no longer crashing! The app launches fine.

I let some audio play for a while and it does still start buzzing after a while. It did take longer though (2-3 minutes vs less than 1 before), and this time it was very abrupt and more like buzzing than the weird warbling before.

I tried switching between the speakers and eqMac in the status bar and then it finally crashed. It opens up again just fine, though, so that's a plus.

**Comment by kesarion** [comment]
It's too bad I have only 16 and 24 bit integer options for my speakers, while eqMac is strictly 32 bit float, so no way of checking if having them in the same format works on my side.

**Comment by nodeful** [comment]
Just curious how many people have tried to decrease the global EQ gain and increasing their overall volume. Like this:
![image](https://user-images.githubusercontent.com/8472525/80644463-97b67780-8a61-11ea-8fe2-21b348042d87.png)
As you can see I'm running a USB DAC with 48kHZ 16bit Int Channel and my sounds is bassy and super clear, no static.
Try to play with global gain, if you have "fixed" your static noise that means it was never broken, just a hard limit of your Audio Hardware.
This probably won't solve most people's issues but some for sure.

**Comment by MrSpethial** [comment]
[Eqmac_Error_On_Installation.txt](https://github.com/bitgapp/eqMac/files/4555835/Eqmac_Error_On_Installation.txt)



> Whoever has problems running v0.1.0 please download the pre-release of v0.1.1 here: https://github.com/bitgapp/eqMac/releases/tag/v0.1.1

I got the attached error after installing (Same issue no sound).
Steps:
1. Uninstalled using the app uninstall.
2. Installed the new version (attached error)
3. No sound when the app is open (Even when disabling the on/off button)

**Comment by rome3ro** [comment]
> Just curious how many people have tried to decrease the global EQ gain and increasing their overall volume. Like this:
> ![image](https://user-images.githubusercontent.com/8472525/80644463-97b67780-8a61-11ea-8fe2-21b348042d87.png)
> As you can see I'm running a USB DAC with 48kHZ 16bit Int Channel and my sounds is bassy and super clear, no static.
> Try to play with global gain, if you have "fixed" your static noise that means it was never broken, just a hard limit of your Audio Hardware.
> This probably won't solve most people's issues but some for sure.

@nodeful How did you get 48 working because on my environment switching to 48 instead 44 when I open the app is just muted!

**Comment by mwmrshll** [comment]
I installed the new version. With my same setup I described before, I had the issue arise again, perhaps more abruptly.

**Comment by migdea11** [comment]
I have been running into the same issue, but when I installed the fix mentioned in this thread the issue went away. [https://github.com/bitgapp/eqMac/issues/180](url) Hopefully this helps identify the issue because, I can't install the 3rd party fix on my work laptop, since it not from a known developer.

**Comment by JohnStarich** [comment]
I see this one's particularly hard to pin down. Hopefully my below hardware config + data will help a bit 🙏 

eqMac: `UI Version: 0.1.0      App Version: 0.1.1`
macOS: `Catalina 10.15.4 (19E287)`
machine: `MacBook Pro (15-inch, 2016)`
audio setup:
```
TB3 dock with headphone jack
format: 2ch 16-bit Integer 48.0 kHz  <>

eqMac
format: 48,000 Hz  <>  2ch 32-bit Float
```

Something I noticed was the regularity in my speakers beginning to buzz. I timed it for several runs, where each time I clicked the audio output device (plugable audio out) to reset the buzz. If it helps, I was streaming music the whole time.

Turns out, the times were pretty consistent, around `57m`:
```
57m34s
58m2s
55m39s
```
Perhaps there's some state that loses precision around 1hr of audio into use?

**Comment by nadstaky** [comment]
Hello Roman,

Installed latest version 0.1.1. (macOS Catalina 10.15.4, 2019 15" MBP with Core i9, 16 GB RAM).
Sound still crackles and distorts when playing through my USB DAC, then no audio at all (Advance Paris Smart DX1).

Please fix this as soon as possible, as I can't test v0.1.1, it's unusable...

Best,
Eric

**Comment by jeffconsi** [comment]
Hello Roman,

Quickly, thank you so much for your work from us users with hearing impairments. I am enjoying music even more now.

I am also having this crackling to no-sound issue using a USB DAC. Happened twice so far after 20-30mins of playback. Clicking the "Reset playback" button fixes the issue. I'm removing the DAC and testing the MacBook soundcard directly now. Will write back if any issues with that. Thanks again.

**Comment by jeffconsi** [comment]
Hi Roman, this issue persists with the internal Mac soundcard as well. Crackling static to no sound at all within about 20-30mins of playback. Reset playback fixes the issue, but I guess it still needs to be addressed.

**Comment by JohnStarich** [comment]
For what it's worth, the buzz seems to start at a specific time into playback. It doesn't appear to gradually increase throughout, just kicks in at about 57m, then gradually increases until music is drowned out. 

Is there a way I can collect these logs to help diagnose?
https://github.com/bitgapp/eqMac/blob/fda5238c544785bcee1098727478970e14297aa6/native/app/Source/Audio/Outputs/Output.swift#L102
I looked through Console.app, but I'm not seeing this particular line for `process:eqMac`

**Comment by jeffstampes** [comment]
> Quickly, thank you so much for your work from us users with hearing impairments. I am
> enjoying music even more now.

And in addition, there are co-workers I can't understand over videoconference because they have such low voices and mics with poor high end.  I have programs now with their names, and I flip over when they start talking.   So it makes me more productive as well.

Thanks for your effort on this issue!!

**Comment by austinbutler** [comment]
Another data point here: AudioQuest DragonFly Black, by default was running at 96,000 Hz, 24-bit integer and sound starts to distort within 30s. Eventually drops out completely. Tried setting both the DAC and eqMac to 44.1, same issue pretty much. macOS 10.15.4.

**Comment by fbcd4f50ea5350f051450323b5abfd0b** [comment]
Same here with a Dragonfly Red on 44.1.
I've created a short video with the noise. I can share it if that's of any help.

**Comment by brandonkboswell** [comment]
I am also getting the same on a 15in, 2018 Macbook running Catalina 10.15.4 into a Schiit Fulla

**Comment by jeffconsi** [comment]
Hi. I downloaded your latest version 0.2.0 and this issue still persists. I love the updates to the UI and the sound is still excellent. Just after 20-30mins it starts to distort and gets worse until it just cuts out altogether. I have a fairly new and powerful machine, so not sure why this is happening.

**Comment by IDAK125** [comment]
Hello. I ran into the same static/choppy sound with 0.2.0 (as with the previous release). However, I find it happens within almost half the time now. With v0.1.1, it was averaging around 20-25 minutes. With 0.2.0 it's averaging around 10-12 minutes. In addition, I find the timing still occurs with or without sound going through the outputs during that time. After enough time, all sound is unresponsive (including system sounds)

Here are the stats:
MacMini 8,1 - i7, 32GB RAM
Audio through the built-in headphone jack
Audio MIDI Setup: 44.1 on both eqMac, Headphone, Built-in Speaker.

Hope this helps.

**Comment by jeffconsi** [comment]
Yes. @IDAK125 I am experiencing the same issue too. If I leave the machine and come back, all sound is unresponsive including system sounds. @nodeful is this something you are finding too?

**Comment by eShini** [comment]
> Hello. I ran into the same static/choppy sound with 0.2.0 (as with the previous release). However, I find it happens within almost half the time now. With v0.1.1, it was averaging around 20-25 minutes. With 0.2.0 it's averaging around 10-12 minutes. In addition, I find the timing still occurs with or without sound going through the outputs during that time. After enough time, all sound is unresponsive (including system sounds)
> 
> Here are the stats:
> MacMini 8,1 - i7, 32GB RAM
> Audio through the built-in headphone jack
> Audio MIDI Setup: 44.1 on both eqMac, Headphone, Built-in Speaker.
> 
> Hope this helps.

I have exactly the same issue here, it always occurs around every 15 minutes, even when no audio is running on the device. It forces me to quit and run again eqmac to fix the static noise for 15 more minutes.
Hope it'll be fixed soon.

**Comment by nodeful** [comment]
Hey everyone, I know this bug is very annoying for a lot of you but again I cannot debug the issue as it never happened to me yet. The difference in macOS versions and Hardware setups will make this one hard to debug. As eqMac works for 90%-95% of the users without this issue I'm focusing on other things at the minute that will make eqMac more sustainable long term. 
When the time comes I will definitely fix this issue as it seems very important for a good user experience for everyone.
Btw, you don't have to restart the app to fix this, I've added a Reset Playback button to the left of the Output Device selection box, just click that and it should mitigate the issue for some time.
Thank you all!

**Comment by nodeful** [comment]
Maybe some other developer comes along like @ekarlo or somebody as brilliant and will fix the issue! I will try and setup an Issue Bounty for this one!

**Comment by IssueHuntBot** [comment]
[@nodeful](https://issuehunt.io/u/nodeful) has funded $100.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/98095061/issues/224) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/98095061/issues/224).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by JohnStarich** [comment]
@nodeful Thanks for your update!

Would you mind pointing me to how to access eqMac’s logs like in this comment? https://github.com/bitgapp/eqMac/issues/224#issuecomment-627104245

It could be very useful in helping us dig deeper into the underlying cause.

**Comment by nodeful** [comment]
@JohnStarich yeah those console logs only get logged during Debug runtime environment. So one would have to run the app from Xcode. I will improve the developer documentation shortly!

**Comment by nodeful** [comment]
I've improved the dev documentation [in the README](https://github.com/bitgapp/eqMac/blob/master/README.md#development) but would have to spend some more time explaining how the code is structured. I will do this at a later date, but if you have any questions regarding development ask on https://discord.eqmac.app

**Comment by darrenwinslow** [comment]
I also am having this issue.  It probably takes an hour or two before this happens (distortion).  If I do a reset it will go away for a while.  Running at 48kHz using an eternal USB audio device.

**Comment by fbcd4f50ea5350f051450323b5abfd0b** [comment]
For me it always takes just a few seconds.

**Comment by IvanTurgenev** [comment]
macbook air retina vanilla catalina affected headphones

**Comment by darrenwinslow** [comment]
Note that changing volume or disabling the eq (slide switch to the left) has no effect on the static.  Also, running within Xcode produced no additional output when this happens.  Also, it does take about 2 hours for the static to become audible, and it gradually gets worse over time to the point when it is just static.

**Comment by artem-kobets-qa** [comment]
Got similar issue.

Here is my configuration:
![image](https://user-images.githubusercontent.com/62426660/84375427-36e59780-abe8-11ea-8047-539d699c6fe3.png)
![image](https://user-images.githubusercontent.com/62426660/84375458-44028680-abe8-11ea-932e-f643a82913ea.png)
<img width="587" alt="Screenshot 2020-06-11 at 13 35 01" src="https://user-images.githubusercontent.com/62426660/84375713-a6f41d80-abe8-11ea-9b4c-bbbc830ece9c.png">

**Comment by darrenw2112** [comment]
@nodeful, I figured out and have a fix for this issue.  I am not a Swift guy, so you will probably have to remove some semicolons.  I should have a chance in the next few days to upload it.  How would you like me to post it (I have no idea what IssueHunt is)?

**Comment by AcidSlide** [comment]
I would like to add my config to the list, I've encountered this even just having/using the built-in speakers.. (using eq 0.2.2), this somehow didn't happen when I was using "0.2.0" version (or might not have noticed it).

Also happened while using an external headphone.

![image](https://user-images.githubusercontent.com/6833047/84614431-ab764a00-aef8-11ea-85ea-705bd42ef452.png)

**Comment by nodeful** [comment]
@darrenw2112 that's awesome man!
Just go here: https://issuehunt.io/r/bitgapp/eqMac/issues/224 (login with GitHub)
And click Submit Pull Request on the right column!

**Comment by jeffconsi** [comment]
Downloaded 0.3 today. Sounds great thank you. But, this issue is still happening only this time the sound gets slightly distorted but doesn't cut out. It just maintains a crackly sound. Clicking 'Reset Playback' fixes the issue, but it comes back after 15-20mins of continuous play. Also, the Balance L/R fader/slider doesn't do anything.

**Comment by AcidSlide** [comment]
I can confirm that the crackling sound still happens after some time.. I haven't logged how long yet.. but issue is still there.

Clicking the reset/resync button fixes the issue temporarily.

By the way, i'm already using eqMac 0.3.1 with 1.1.0 drivers.

**Comment by darrenw2112** [comment]
@nodeful I will try to post this today.

**Comment by darrenw2112** [comment]
@nodeful I submitted the pull request.  For some reason it shows a lot has changed.  The only changes are in:
CARubuffer.swift -> clipTimeBounds
Output.swift -> init
Output.swift -> outputRenderCallback
Output.swift -> a few class variables

**Comment by nodeful** [comment]
@darrenw2112 I will take a look this evening, my laptop has died unfortunately waiting for a replacement at the moment so wont be able to test for a little bit

**Comment by nodeful** [comment]
Hey everyone 👋 
I have created a pre-release candidate with the fixes by @darrenw2112 that will potentially fix this issue!
Please download it here: https://github.com/bitgapp/eqMac/releases/tag/v0.3.2
And report back if it solved this issue for you

**Comment by darrenw2112** [comment]
Note that I changed your init back to the way you had it because the rates that were reported in your new way (using the mix bus) did not reflect what the input was set to.  It always reported 44.1.

**Comment by AcidSlide** [comment]
> Hey everyone 👋
> I have created a pre-release candidate with the fixes by @darrenw2112 that will potentially fix this issue!
> Please download it here: https://github.com/bitgapp/eqMac/releases/tag/v0.3.2
> And report back if it solved this issue for you

I'm currently running in 25+ mins already and so far I haven't noticed any crackling or static noises.  I will do another update at around 1+ hr mark.

Checking CPU usage, Activity Monitor is reporting an average of 3.4% with average of 15 threads running (main process only).
![image](https://user-images.githubusercontent.com/6833047/85632594-2e8e6180-b6aa-11ea-99a9-5bb0edcae896.png)
I'm just not sure why CPU Time is reporting more than 1hr.  I just installed 0.3.2 about 30mins ago.

**Comment by AcidSlide** [comment]
> Hey everyone 👋
> I have created a pre-release candidate with the fixes by @darrenw2112 that will potentially fix this issue!
> Please download it here: https://github.com/bitgapp/eqMac/releases/tag/v0.3.2
> And report back if it solved this issue for you

Ok after doing some additional checking the Activity Monitor, I noticed CPU usage on 2 out my 16 hyperthread cores was at more than 10% so I went to find it. (NOTE: No music/audio is playing while I was checking)

1st... "coreaudiod" was spiking between 10% and 20%, is this normal??
![image](https://user-images.githubusercontent.com/6833047/85633630-33ecab80-b6ac-11ea-8e82-b20b0e0d9b2e.png)


2nd... After quitting eqMac "coreaudiod" went back to 0-0.2% usage and at the same time the audio driver was not removed from the enabled audio devices
![image](https://user-images.githubusercontent.com/6833047/85633741-6e564880-b6ac-11ea-9c2c-ea5ef68072e9.png)

**Comment by darrenw2112** [comment]
I have been running for several hours and coreaudiod < 3% with eqMac around 5% (same as in Xcode).

**Comment by AcidSlide** [comment]
> I have been running for several hours and coreaudiod < 3% with eqMac around 5% (same as in Xcode).

Hmmm mine is at 10-11% average (for coreaudiod) CPU usage with or without music playing.

MBP Processor: 2.3Ghz 8-Core Core i9

What's your processor?

**Comment by AcidSlide** [comment]
By the way, I'm already at 45mins mark with music playing via Spotify and sound is via MacBook Pro Speaker..

So far, no crackling or static noises heard.. audio quality is still at 100% for me..

**Comment by darrenw2112** [comment]
MBP:  2017 -> 2.9 Ghz Quad-Core i7. 
coreaudiod is a fickle one

**Comment by nodeful** [comment]
Currently the way eqMac works is it processes audio all the time, even if it's silence, there's no real nice way to fix that without going absolutely crazy in to optimisation, which is not something I would like to focus on right now personally. For me on my 2.3 GHz Dual-Core Intel Core i5 eqMac together with CoreAudio uses around 5-7% CPU on average which is totally acceptable for me. You have to take into consideration other apps that might be interacting with CoreAudio in any way, even if it's not apparent straight away. Also higher Sampling Rate devices and more complex protocols like Thunderbolt can increase the load on CoreAudio in unknown ways. Regarding this issue, let's see what some more people will report, but so far looks very promising. Congrats and thanks @darrenw2112 couldn't do it without you. Would be nice to chat with you in https://discord.eqmac.app to understand how you fixed the problem in more detail :)

**Comment by AcidSlide** [comment]
> Currently the way eqMac works is it processes audio all the time, even if it's silence, there's no real nice way to fix that without going absolutely crazy in to optimisation, which is not something I would like to focus on right now personally. For me on my 2.3 GHz Dual-Core Intel Core i5 eqMac together with CoreAudio uses around 5-7% CPU on average which is totally acceptable for me. You have to take into consideration other apps that might be interacting with CoreAudio in any way, even if it's not apparent straight away. Also higher Sampling Rate devices and more complex protocols like Thunderbolt can increase the load on CoreAudio in unknown ways. Regarding this issue, let's see what some more people will report, but so far looks very promising. Congrats and thanks @darrenw2112 couldn't do it without you. Would be nice to chat with you in https://discord.eqmac.app to understand how you fixed the problem in more detail :)

Thanks for the update, I do understand the possible intricacies of coreaudiod.  If I can find extra time next or next next month, I might try to help out on the optimizations. 😁

Thanks @nodeful for creating such a great app and thanks to @darrenw2112 for finding a fix to the audio crackling/noise.

For running at more than 1hr mark, audio quality still at 100% 😊

**Comment by AcidSlide** [comment]
> > Hey everyone 👋
> > I have created a pre-release candidate with the fixes by @darrenw2112 that will potentially fix this issue!
> > Please download it here: https://github.com/bitgapp/eqMac/releases/tag/v0.3.2
> > And report back if it solved this issue for you
> 
> I'm currently running in 25+ mins already and so far I haven't noticed any crackling or static noises. I will do another update at around 1+ hr mark.
> 
> Checking CPU usage, Activity Monitor is reporting an average of 3.5% with average of 15 threads running (main process only).
> ![image](https://user-images.githubusercontent.com/6833047/85632594-2e8e6180-b6aa-11ea-99a9-5bb0edcae896.png)
> I'm just not sure why CPU Time is reporting more than 1hr. I just installed 0.3.2 about 30mins ago (previously uninstalled 0.3.1 yesterday).

Oh no! I think a new bug was introduced... no sound coming out of my external port using a headphones, no sound coming out.

It also crashed while trying to switch between "MacBookPro Speaker" and "External Headphones"

NOTE: But past 2hr mark, still no crackling or static noises observed.

**Comment by HerrTamm** [comment]
For me static noise always started after 5 seconds and sound was a total robot mess in just 10 seconds. Updated app to 0.3.2 and driver to 1.1.1 and so far so good, no problems anymore. CPU usage is 2-4% (i7 8750H)

**Comment by darrenw2112** [comment]
I have not used discord before (here is my excuse). A chat would be good.

**Comment by darrenw2112** [comment]
I occasionally have a crash when switching inputs (I believe this is in the init function).  Also it would be great if the program would not switch the output device when a new one becomes available.

**Comment by HerrTamm** [comment]
> I occasionally (seems to happen much less anymore) have a crash when switching inputs (I believe this is in the init function). Also it would be great if the program would not switch the output device when a new one becomes available.

You should probably create a new issue or find a more relevant one: https://github.com/bitgapp/eqMac/issues

**Comment by darrenw2112** [comment]
I know, I was just responding to the comment earlier about it stopping.

**Comment by Xardas220** [comment]
Hi, I installed the new Version and driver and what should I say - it's gone. The crackling is gone completely. I listen now 3h to music, YouTube and other things and all works as aspected. Thank you, @darrenw2112 for the fix and @nodeful for the implementation. :)
I compared the CPU usage before and after and saw no difference. I am on a hackintosh (R7 3700X, 16GB, 5700XT), so plenty of horsepower. The usage was before and after between 0.8% and 1.4%. 15 Threads running in within the eqMac process. Coreaudiod is pinned on 0.7% for the whole time, running 14 Threads. Nothing special in summery.

**Comment by nodeful** [comment]
Thanks for the feedback everyone. I made the v0.3.2 release public now. Seems like this specify issue can now be finally closed.

**Comment by Hairpin2020** [comment]
I've been using eqMac for the last few years and I've never had any issues with it up until very recently. 

I'm having this issue on Mac OS High Sierra - the audio starts crackling after about an hour of using eqMac and it would only resolve itself if I quit and restart the app.

I had to recently reset my PRAM and eqMac is completely unusable (noise crackling when changing the volume up or down, bass is constantly crackling and the Advanced equalizer makes everything worse.

Mac OS Audio devices Format 48,000 Hz.
eqMac version 0.3.7
UI version 1.2.2
Driver version: 1.2
